src/irmin-pack/IO_intf.ml

```diff
@@ -16,27 +16,85 @@
 
 open! Import
 
+(** This interface defines a type [t] which wraps a file descriptor and provides some
+    additional support for metadata fields (version, offset), and support for periodic
+    flushing of data (writes to file go via an internal buffer which is flushed when the
+    amount of buffered data is beyond a certain limit). *)
 module type S = sig
   type t
+
   type path := string
 
   val v : version:Version.t option -> fresh:bool -> readonly:bool -> path -> t
+  (** [v ~version ~fresh ~readonly pth] creates a new IO instance. [version] must be
+      [Some] if the [pth] does not point to a file. If the file exists, then [fresh] will
+      cause all data to be discarded (except for metadata); if [fresh] is false, and the
+      version of the file is greater than [version], an exception is thrown. [readonly]
+      indicates that operations like {!append} which change the file data or metadata
+      should be disallowed. 
+
+      In order to store the metadata, a [header_size] of 16 bytes is reserved at the
+      beginning of the file; reads and writes to the file at offset [off] will be
+      translated to offset [off+header_size].  
+
+      For a given [pth], there should be at most one read/write instance. There can be
+      multiple readonly instances, but only one read/write instance is allowed. This is
+      not currently enforced (but could be, using lock files for example).  *)
+
   val name : t -> string
+  (** [name t] is the [path] argument provided to {!v} when [t] was created *)  
+
   val append : t -> string -> unit
+  (** [append t s] appends [s] to the end of the file pointed to by [t]. The data is first
+      placed in a buffer, and a flush to file only happens when the amount of data in the
+      buffer exceeds the [auto_flush_limit], which is currently 1M.
+
+      [append] and [set] are the only operations that write to the file, apart from
+      metadata operations. *)
+
   val set : t -> off:int63 -> string -> unit
+  (** [set t ~off s] writes the string [s] at offset [off+header_size]. *)
+
   val read : t -> off:int63 -> bytes -> int
+  (** [read t ~off buf] tries to read [Bytes.length buf] bytes from [t] at offset [off +
+      header_size]; it returns the number of bytes actually read, which may be less than
+      requested if we are near the end of the file *)
+
   val read_buffer : t -> off:int63 -> buf:bytes -> len:int -> int
+  (** [read_buffer t ~off ~buf ~len] tries to read [len] bytes from [t] at offset
+      [off+header_size] *)
+
   val offset : t -> int63
```

Comment:
Ioana: clarify what type of offset (real vs offset-in-data); don't talk about append